src/app/mesh/qgsmeshdatasetgrouptreewidget.cpp

```diff
@@ -0,0 +1,109 @@
+  connect( mExpandButton, &QPushButton::clicked, mDatasetGroupTreeView, &QTreeView::expandAll );
+  connect( mCheckAllButton, &QPushButton::clicked, mDatasetGroupTreeView, &QgsMeshDatasetGroupTreeView::selectAllGroups );
+  connect( mUnCheckAllButton, &QPushButton::clicked, mDatasetGroupTreeView, &QgsMeshDatasetGroupTreeView::deselectAllGroups );
+  connect( mResetDefaultButton, &QPushButton::clicked, this, [this]
+  {
+    this->mDatasetGroupTreeView->resetDefault( this->mMeshLayer );
+  } );
+
+}
+
+void QgsMeshDatasetGroupTreeWidget::syncToLayer( QgsMeshLayer *meshLayer )
+{
+  mMeshLayer = meshLayer;
+  mDatasetGroupTreeView->syncToLayer( meshLayer );
+}
+
+void QgsMeshDatasetGroupTreeWidget::apply()
+{
+  if ( mMeshLayer )
+    mMeshLayer->setDatasetGroupTreeRootItem( mDatasetGroupTreeView->datasetGroupTreeRootItem() );
+}
+
+void QgsMeshDatasetGroupTreeWidget::addDataset()
+{
+  if ( !mMeshLayer->dataProvider() )
+    return;
+
+  QgsSettings settings;
+  QString openFileDir = settings.value( QStringLiteral( "lastMeshDatasetDir" ), QDir::homePath(), QgsSettings::App ).toString();
+  QString openFileString = QFileDialog::getOpenFileName( nullptr,
+                           tr( "Load mesh datasets" ),
+                           openFileDir,
+                           QgsProviderRegistry::instance()->fileMeshDatasetFilters() );
+
+  if ( openFileString.isEmpty() )
+  {
+    return; // canceled by the user
+  }
+
+  QFileInfo openFileInfo( openFileString );
+  settings.setValue( QStringLiteral( "lastMeshDatasetDir" ), openFileInfo.absolutePath(), QgsSettings::App );
+  QFile datasetFile( openFileString );
+
+  bool isTemporalBefore = mMeshLayer->dataProvider()->temporalCapabilities()->hasTemporalCapabilities();
+  bool ok = mMeshLayer->dataProvider()->addDataset( openFileString );
+  if ( ok )
+  {
+    QgsMeshLayerTemporalProperties *temporalProperties = qobject_cast< QgsMeshLayerTemporalProperties * >( mMeshLayer->temporalProperties() );
+    if ( !isTemporalBefore && mMeshLayer->dataProvider()->temporalCapabilities()->hasTemporalCapabilities() )
+    {
+      mMeshLayer->temporalProperties()->setDefaultsFromDataProviderTemporalCapabilities(
+        mMeshLayer->dataProvider()->temporalCapabilities() );
+
+      if ( ! temporalProperties->referenceTime().isValid() )
+      {
+        QDateTime referenceTime = QgsProject::instance()->timeSettings()->temporalRange().begin();
+        if ( !referenceTime.isValid() ) // If project reference time is invalid, use current date
+          referenceTime = QDateTime( QDate::currentDate(), QTime( 0, 0, 0, Qt::UTC ) );
+        temporalProperties->setReferenceTime( referenceTime, mMeshLayer->dataProvider()->temporalCapabilities() );
+      }
+
+      mMeshLayer->temporalProperties()->setIsActive( true );
+    }
+    QMessageBox::information( this, tr( "Load mesh datasets" ), tr( "Datasets successfully added to the mesh layer" ) );
+    emit mMeshLayer->dataSourceChanged();
```

Comment:
It is not generally a good practice to emit signals on behalf of other objects...

What do you think of moving this chunk of code (lines 80-101, without QMessageBox) to a new method in QgsMeshLayer?